File: airesources/Java/hlt/Navigation.java
package hlt;

public class Navigation {

    private Ship ship;
    private Entity target;

    public Navigation(final Ship ship, final Entity target) {
        this.ship = ship;
        this.target = target;
    }

    public ThrustMove navigateToDock(final GameMap gameMap, final double maxThrust) {
        final int maxCorrections = Constants.MAX_CORRECTIONS;
        final boolean avoidObstacles = true;
        final int angularStep = 1;
        final Position targetPos = ship.getClosestPoint(target);

        return navigateTowards(gameMap, targetPos, maxThrust, avoidObstacles, maxCorrections, angularStep);
    }

    public ThrustMove navigateTowards(final GameMap gameMap, final Position targetPos, final double maxThrust,
                                      final boolean avoidObstacles, final int maxCorrections, final int angularStep) {
        if (maxCorrections <= 0) {
            return null;
        }

        final double distance = ship.getDistanceTo(targetPos);
        final double angleDeg = ship.orientTowardsInDeg(targetPos);

        if (avoidObstacles && !gameMap.objectsBetween(ship, targetPos).isEmpty()) {
            final double newTargetDx = Math.cos(Math.toRadians(angleDeg + angularStep)) * distance;
            final double newTargetDy = Math.sin(Math.toRadians(angleDeg + angularStep)) * distance;
            final Position newTarget = new Position(ship.getXPos() + newTargetDx, ship.getYPos() + newTargetDy);

            return navigateTowards(gameMap, newTarget, maxThrust, true, (maxCorrections - 1), angularStep);
        }

        final double thrust;
        if (distance < maxThrust) {
            thrust = distance;
        }
        else {
            thrust = maxThrust;
        }

        return new ThrustMove(ship, (int)(angleDeg + 0.5), (int) (thrust + 0.5));
    }
}

File: airesources/Scala/hlt/Navigation.scala
package hlt

import scala.annotation.tailrec

class Navigation(var ship: Ship, var target: Entity) {
  def navigateToDock(gameMap: GameMap, maxThrust: Int): Option[ThrustMove] = {
    val maxCorrections = Constants.MAX_CORRECTIONS
    val avoidObstacles = true
    val angularStepRad = Math.PI / 180
    val targetPos = ship.getClosestPoint(target)
    navigateTowards(gameMap, targetPos, maxThrust, avoidObstacles, maxCorrections, angularStepRad)
  }

  @tailrec
  private def navigateTowards(gameMap: GameMap,
                              targetPos: Position,
                              maxThrust: Int,
                              avoidObstacles: Boolean,
                              maxCorrections: Int,
                              angularStepRad: Double): Option[ThrustMove] = {
    if (maxCorrections <= 0) {
      None
    } else {
      val distance = ship.getDistanceTo(targetPos)
      val angleRad = ship.orientTowardsInRad(targetPos)
      if (avoidObstacles && !gameMap.objectsBetween(ship, targetPos).isEmpty) {
        val newTargetDx = Math.cos(angleRad + angularStepRad) * distance
        val newTargetDy = Math.sin(angleRad + angularStepRad) * distance
        val newTarget = new Position(ship.getXPos + newTargetDx, ship.getYPos + newTargetDy)
        navigateTowards(gameMap,
                        newTarget,
                        maxThrust,
                        avoidObstacles = true,
                        maxCorrections - 1,
                        angularStepRad)
      } else {
        var thrust = 0
        if (distance < maxThrust) { // Do not round up, since overshooting might cause collision.
          thrust = distance.toInt
        } else {
          thrust = maxThrust
        }
        val angleDeg = Util.angleRadToDegClipped(angleRad)
        Some(new ThrustMove(ship, angleDeg, thrust))
      }
    }
  }
}
